joomla-extensions/patchtester
# [4.0] Patch tester enhancement for joomla4 (pre-compiled diff)

## Patchtester enhancement for Joomla4

Until Joomla 4, Joomla did not have any assets that needed to be comiled in any way. Therefore, testing pull requests could be made a little more easy for people who are not into dev-stuff and do not have installed any dev-tools (except xampp/MAMP/LAMP... maybe). The patchtester called the Github API and asked for any changes, downloaded those changed files and exchanged it on the client machines (note that it did **not** only apply the diff; therefore a git client would be needed). 

In Joomla 4 with the new front-end, you need to compile the source files with composer and npm before you are good to go with your Joomla instance. As many users don't know how to use these tools, we [provide a pre-compiled version of Joomla](https://developer.joomla.org/nightly-builds.html) which the user has to unzip and deploy on his server. There are no tools needed except a (s)ftp client or xampp/MAMP/LAMP...! 

Applying "patches" the way patchtester works is possible but very limited. You can't get any compiled files from Github, so you only apply the source files, but not the compiled ones Joomla uses. For this very reason, we came to the conclusion that pre-compiling and deploying this pre-compiled diff is the best solution for now to be able to keep serving the patchtester for our users.

Therefore, we already implemented another step in the Drone CI automation that will search for any diffs in the compiled version of the newest Joomla 4.0-dev-Branch with the branch that the pull request is based on, as you can see in this [pull request joomla/joomla-cms: #26274](https://github.com/joomla/joomla-cms/pull/26247). This step is done by a new, fairly lightweight docker image which is now included in the joomla-projects/docker-images repository.

For the patchtester itself, the following changes were made:

### Summary of Changes

#### Changes visible to the user

* A switch is added where the user can choose where the diff should come from. Per default, since this enhancement is developed for the purpose of testing Joomla 4, the CI server is selected. 

* You can specify the location of the CI server in the global settings.

* Since pull requests can build on each other and the patchtester is able to apply multiple pull request "patches", the user is forced to remove the patches in the exact opposite order he applied them. For usability reasons, he is still able to use the reset button on the top of the patchtester to revert all changes. The patchtester will then automatically revert the patches in the correct order, no matter where the changes came from (Github/CI).

* Language strings for `en-gb`, `en-us` and `de-de` were added for the CI mechanism.

#### Changes inside the "black box"

* Applying and reverting changes was refactored so that different routines are executed when called, depending on whether the Github or CI technique was selected in global configuration. 

* For the CI mechanism, new methods to fetch, apply and revert changes where implemented.

* When applying multiple patches via CI, a patch chain was added which keeps which pull requests were added and in which order. The patch chain is stored in a new table `#__patchtester_chain`. This is necassary to revert the changes correctly without breaking Joomla, which has been an issue on patchtester so far. The patch chain is not used when using Github so far, but could be implemented if wanted. 

* The buttons where refactored to event-handling (no more inline-href-javascript).

* The old template overwrites were removed.

### Testing Instructions

* Get a fresh Joomla 4 installation. You can get Joomla 4 without any tooling needed here: [Joomla Nightly Builds](https://developer.joomla.org/nightly-builds.html)

* Download this PR as zip from this pull request, build the extension using the build script and install it as you would install any other extension in Joomla.

* Search for a pull request which already has a diff build and deployed by Drone CI on the CI server.

* In the backend of Joomla, go to patchtester, update the pull requests using the fetch data button and click on apply at your selected pull request.

* Check if the changes proposed by the selected pull requests where applied.

* Go to `https://ci.joomla.org:444/artifacts/joomla/joomla-cms/4.0-dev/<pull-request-id>/patchtester/build.zip` and download the zip file with your browser after you inserted the pull request id of you pull request in the corresponding place of the url.

* Open up that zip file. It should contain all new or changed files of the pull request as well as a list of deleted files if there are any in the pull request.

* Check your joomla installation for those files and their file contents.

* Revert the changes using either the revert button in patchtester or by clicking on reset.

* Check your joomla installation again. The original files should be recovered and no files of the applied changes should be there. 

* **Please feel free to think outside the box when testing this!**
